torch_geometric/loader/cache.py

```diff
@@ -0,0 +1,55 @@
+from typing import Any, List, Tuple
+
+import torch
+from torch import Tensor
+from torch.utils.data import DataLoader
+
+BatchType = List[Tuple[Tensor, Tensor, int]]
+
+
+class CachedLoader:
+    r"""A caching loader intended to use in layer-wise inference
+    performed on GPU. During the first walk through the data
+    stored in :class:`torch.utils.data.DataLoader` batches are
+    transferred to GPU memory and cached.
+
+    Args:
+        loader (torch.utils.data.DataLoader): The wrapped data loader.
+        device (torch.device): The device to load the data to.
+    """
+    def __init__(self, loader: DataLoader, device: torch.device):
+        self.n_batches = len(loader)
+        self.iter = iter(loader)
+        self.cache_filled = False
+        self.device = device
+        self.cache = []
+
+    def cache_batch(self, batch: Any) -> BatchType:
```

Comment:
I'm curious about your thoughts around the interface.

As a user who has a different hardware requirement and dataset, I may want to modify which attributes to cache. I wonder what a clean interface looks like to allow this.

Do you think it's feasible to always return a data object from `__call__`? Currently, I don't think it's very intuitive for users because `next(iter(CachedLoader(loader)))` is a tuple of some of attributes while `next(iter(loader))` is a data object.

Reply:
Thanks for raising really good points.

> As a user who has a different hardware requirement and dataset, I may want to modify which attributes to cache. I wonder what a clean interface looks like to allow this.

I changed the design a bit, now user can register attributes hooks via `register_attr_hook`. It requires a callable object that accepts batch and returns attribute.

> Do you think it's feasible to always return a data object from __call__? Currently, I don't think it's very intuitive for users because next(iter(CachedLoader(loader))) is a tuple of some of attributes while next(iter(loader)) is a data object.

Unfortunately, I don't think it is possible, it will require caching batches, which will move us back to a problem with increased memory usage (described [here](https://github.com/pyg-team/pytorch_geometric/pull/7896#discussion_r1297570176)). I have modified the documentation a bit, hopefully, it will clear some of user confusion.